### DISCUSSION TITLE
BluetoothA2DPSink and MemoryStream of a byte[] sound both to I2sStream

### DISCUSSION THREAD

**Discussion by vincenzolanera** [pschatzmann/arduino-audio-tools#1676] [Q&A] [answered]
Hi all,  I'm trying to build a bluetooth speaker.
I'm using **BluetoothA2DPSink** and **I2SStream** for sending data to a **PCM5102 DAC**.

I would play a sound (eg. a voice that tells "Connected") when a bluetooth device is connected.

It works only at the first time. If I turn off and on the bluetooth of my phone, the second time the voice doesn't works anymore and the followin error is logged in the serial monitor:

``[E] StreamCopy.h : 401 - write  to target has failed after 22 retries! (1024 bytes)``

**What is the proper way to reproduce the sound inside connected callback?**

```
I2SStream i2sStream;
BluetoothA2DPSink bluetooth_sink(i2sStream);

MemoryStream connectedSoundStream(connected_sound_bytes, connected_sound_len);
StreamCopy connectedSoundCopyier(i2sStream, connectedSoundStream);
...
void setupBluetooth() {
  auto cfg = i2sStream.defaultConfig();
  cfg.pin_bck = I2S_OUT_BCK_PIN;
  cfg.pin_ws = I2S_OUT_WS_PIN;
  cfg.pin_data = I2S_OUT_DATA_PIN;
  i2sStream.begin(cfg);

  bluetooth_sink.set_avrc_connection_state_callback(avrc_connection_state_callback);
  bluetooth_sink.set_avrc_metadata_callback(avrc_metadata_callback);
  //bluetooth_sink.set_avrc_metadata_attribute_mask(ESP_AVRC_MD_ATTR_TITLE | ESP_AVRC_MD_ATTR_PLAYING_TIME);
  bluetooth_sink.start(BLUETOOTH_NAME);
}
void avrc_connection_state_callback(bool connected) {
  digitalWrite(LED_BUILTIN, connected);
  if(connected) {
    playConnectedVoice();
  } 
}v
void playConnectedVoice() {
  connectedSoundStream.begin();
    while(connectedSoundStream) {
      connectedSoundCopyier.copy();
    }
  connectedSoundStream.end();
}
```

**Comment by pschatzmann** [comment, marked as answer]
If you connect the BluetoothA2DPSink with a I2SStream, the i2s is automatically started and ended by the sink.

You try to write to a closed i2sStream!

So either open it by calling begin() in your playConnectedVoice or don't connect the i2sStream (and thus always leave it open) and do the [a2dp audio output from the callback](https://github.com/pschatzmann/ESP32-A2DP?tab=readme-ov-file#accessing-the-sink-data-stream-with-callbacks) yourself.
